File: ips.py
#!/bin/python3

import re   #used to define patterns to detect sql injection
from scapy.all import sniff, IP, TCP  #scapy a network manipulation tool
#sniff to sniff packets, ip address and TCP for tcp connection
# Signature for SQL Injection (basic example)
sql_injection_pattern = re.compile(r"(\b(select|insert|update|delete|drop|union|into|load_file|outfile)\b.*[';]+)")  #basic regular expression that matches common sql injection such as SELECT INSERT UPDATE DROP

# List of blocked IPs (for demonstration purposes)
blocked_ips = []

def is_sql_injection(payload):
    """
    This function checks if the payload contains SQL Injection patterns.
    ACCEPTS the payloads and checks if it contain sql injection patterns	
    """
    if sql_injection_pattern.search(payload):
        return True
    return False

def block_ip(ip):
    """
    Adds the given IP address to the blocked IP list and prints a message.
    when attack is detected , block_ip functions adds ip to blocked_ips
    """
    if ip not in blocked_ips:
        blocked_ips.append(ip)
        print(f"Blocking IP: {ip}")

def packet_callback(packet):
    """
    This function is called for each captured packet.
    It inspects the packet for SQL injection attempts.
    the core function that processes each captured packet
    the function has an iplayer and a tcp layer
    extract the source and destinantion ip and payload
    """
    if packet.haslayer(IP) and packet.haslayer(TCP):
        ip_src = packet[IP].src
        ip_dst = packet[IP].dst

        # Extract the payload (the data inside the TCP packet)
        payload = str(packet[TCP].payload)

        # If a potential SQL Injection is found in the payload
        if is_sql_injection(payload):
            print(f"Potential SQL Injection detected from {ip_src} to {ip_dst}: {payload}")
            block_ip(ip_src)

def start_sniffing():
    """
    Starts sniffing network packets and applies the callback function.
    """
    print("Starting packet sniffing...")
    sniff(prn=packet_callback, store=0)

if __name__ == "__main__":
    start_sniffing()

File: ips.rs
extern crate pnet;
extern crate regex;

use pnet::datalink::{self, NetworkInterface};
use pnet::packet::{Packet, ipv4::Ipv4Packet, tcp::TcpPacket};
use pnet::util::checksum;
use regex::Regex;
use std::collections::HashSet;
use std::net::Ipv4Addr;
use std::time::Duration;
use std::thread;

// Pattern to match potential SQL injection keywords (simple example)
lazy_static::lazy_static! {
    static ref SQL_INJECTION_PATTERN: Regex = Regex::new(r"(\b(select|insert|update|delete|drop|union|into|load_file|outfile)\b.*[';]+)").unwrap();
}

struct IPS {
    blocked_ips: HashSet<Ipv4Addr>,
}

impl IPS {
    fn new() -> Self {
        IPS {
            blocked_ips: HashSet::new(),
        }
    }

    fn is_sql_injection(&self, payload: &str) -> bool {
        SQL_INJECTION_PATTERN.is_match(payload)
    }

    fn block_ip(&mut self, ip: Ipv4Addr) {
        if !self.blocked_ips.contains(&ip) {
            self.blocked_ips.insert(ip);
            println!("Blocking IP: {}", ip);
        }
    }

    fn start_sniffing(&mut self) {
        let interfaces = datalink::interfaces();
        let iface = interfaces.into_iter().find(|iface: &NetworkInterface| iface.is_up()).expect("No network interface found");

        println!("Starting packet sniffing on interface: {}", iface.name);

        let (mut tx, mut rx) = datalink::channel(&iface, Default::default()).expect("Error opening datalink channel");

        loop {
            let packet = rx.next().expect("Error receiving packet");
            self.process_packet(packet);
        }
    }

    fn process_packet(&mut self, packet: pnet::datalink::DataLinkReceiver) {
        if let Some(ip_packet) = Ipv4Packet::new(packet.packet()) {
            if let Some(tcp_packet) = TcpPacket::new(ip_packet.payload()) {
                let src_ip = ip_packet.get_source();
                let dst_ip = ip_packet.get_destination();
                let payload = String::from_utf8_lossy(tcp_packet.payload());

                if self.is_sql_injection(&payload) {
                    println!("Potential SQL Injection detected from {} to {}: {}", src_ip, dst_ip, payload);
                    self.block_ip(src_ip);
                }
            }
        }
    }
}

fn main() {
    let mut ips = IPS::new();
    ips.start_sniffing();
}

/* enter cargo.toml
[dependencies]
pnet = "0.30.0"  # For packet sniffing
regex = "1.5"     # For pattern matching
*/
